internal/server/filesystem/server_test.go

```diff
@@ -0,0 +1,238 @@
+	return true, nil
+}
+func (fakeFileSystemAPI) Mkdir(path string) error {
+	return nil
+}
+func (fakeFileSystemAPI) Rmdir(path string) error {
+	return nil
+}
+func (fakeFileSystemAPI) LinkPath(tgt string, src string) error {
+	return nil
+}
+
+func TestMkdirWindows(t *testing.T) {
+	v1alpha1, nil := apiversion.NewVersion("v1alpha1")
+	testCases := []struct {
+		name string
+		path string
+		pathCtx	internal.PathContext
+		version apiversion.Version
+		expectError bool
+	}{
+		{
+			name: "absolute path outside of container context with container context set",
+			path: `C:\foo\bar`,
+			pathCtx: internal.CONTAINER,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "absolute path inside container context with container context set",
+			path: `C:\var\lib\kubelet\pods\pv1`,
+			pathCtx: internal.CONTAINER,
+			version: v1alpha1,
+			expectError: false,
+		},
+		{
+			name: "absolute path outside of plugin context with plugin context set",
+			path: `C:\foo\bar`,
+			pathCtx: internal.PLUGIN,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "absolute path inside plugin context with plugin context set",
+			path: `C:\var\lib\kubelet\plugins\pv1`,
+			pathCtx: internal.PLUGIN,
+			version: v1alpha1,
+			expectError: false,
+		},
+		{
+			name: "relative path with invalid character `:` beyond drive letter prefix",
+			path: `csi-plugin\pv1:foo`,
+			pathCtx: internal.PLUGIN,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "relative path with invalid character `/`",
+			path: `csi-plugin\pv1/foo`,
+			pathCtx: internal.CONTAINER,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "relative path with invalid character `*`",
+			path: `csi-plugin\pv1*foo`,
+			pathCtx: internal.PLUGIN,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "relative path with invalid character `?`",
+			path: `csi-plugin?pv1\foo`,
+			pathCtx: internal.CONTAINER,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "relative path with invalid character `|`",
+			path: `csi-plugin|pv1\foo`,
+			pathCtx: internal.PLUGIN,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "relative path with invalid characters `..`",
+			path: `csi-plugin\..\..\..\system32`,
+			pathCtx: internal.CONTAINER,
+			version: v1alpha1,
+			expectError: true,
+		},
+		{
+			name: "absolute path with invalid prefix `\\`",
+			path: `\\csi-plugin\..\..\..\system32`,
+			pathCtx: internal.CONTAINER,
+			version: v1alpha1,
+			expectError: true,
+		},
+	}
+	srv, err := NewServer("windows", `C:\var\lib\kubelet\plugins`, `C:\var\lib\kubelet\pods`, &fakeFileSystemAPI{})
+	if err != nil {
+		t.Fatalf("FileSystem Server could not be initialized for testing: %v", err)
+	}
+	for _, tc := range testCases {
+		t.Logf("test case: %s", tc.name)
+		req := &internal.MkdirRequest {
+			Path: tc.path,
+			Context: tc.pathCtx,
+		}
+		mkdirResponse, _ := srv.Mkdir(context.TODO(), req, tc.version)
```

Comment:
This got me wondering if it might be better for the internal server struct to define their callbacks either with or without (if they don't need it) version argument, and amend the autogen accordingly... Any take on this?

Reply:
Good point. I guess in most scenarios we will not need the version argument. Probably fine to keep it in for now till at least v1beta1 and see?